compiler/rustc_ast/src/ast.rs

```diff
@@ -1536,11 +1536,21 @@ pub enum MacArgs {
     Eq(
         /// Span of the `=` token.
         Span,
-        /// "value" as a nonterminal token.
-        Token,
+        /// The "value".
+        MacArgsEqKind,
     ),
 }
 
+// The RHS of a `MacArgs::Eq` starts out as an expression, either a `Literal`
+// or a `MacCall`. Once macro expansion is completed, all valid cases end up as
+// a literal, which is the form used after lowering to HIR and also for some
+// edge cases where a `MacArgsEqKind` is created from something else.
+#[derive(Clone, Encodable, Decodable, Debug)]
+pub enum MacArgsEqKind {
+    Ast(P<Expr>),
+    Hir(Lit),
```

Comment:
This is an `ast::Lit`, right?
It can be a `token::Lit` instead, less to encode, and closer to the likely future `TokenStream` representation.

Reply:
If we do that, then `MetaItemKind::from_mac_args` must create an `ast::Lit` from a `token::Lit` for the `Hir` case, which is reachable. I think leaving it as an `ast::Lit` is simpler.

Reply:
Ok.